File: Algorithms/Codesignal-Solutions/Codesignal-Arcade/Level 02/adjacentElementsProduct.java
/*
* CODEFIGHTS - ARCADE - LEVEL 02 - ADJACENT ELEMENTS PRODUCT
*
*
* Help source:
* https://stackoverflow.com/questions/17324872/charat-unable-to-extract-individual-members-with-array
*
* Given an array of integers, find the pair of adjacent elements that
* has the largest product and return that product.
*
* Example:
* For inputArray = [3, 6, -2, -5, 7, 3], the output should be
* adjacentElementsProduct(inputArray) = 21.
*
* 7 and 3 produce the largest product.
* Input/Output

[execution time limit] 3 seconds (java)

[input] array.integer inputArray

An array of integers containing at least two elements.

Guaranteed constraints:
2 ≤ inputArray.length ≤ 10,
-1000 ≤ inputArray[i] ≤ 1000.

[output] integer

The largest product of adjacent elements.

[Java] Syntax Tips

// Prints help message to the console
// Returns a string
//
// Globals declared here will cause a compilation error,
// declare variables inside the function instead!
String helloWorld(String name) {
    System.out.println("This prints to the console when you Run Tests");
    return "Hello, " + name;
}
*
*/

public class adjacentElementsProduct {

    static int adjacentElementsProductMethod(int[] inputArray) {

        int product = inputArray[0] * inputArray[1];
        // 3 x 6 = 18

        for(int i = 1; i < inputArray.length-1; i++){

            int temp = inputArray[i] * inputArray[i+1];
            // 6 x -2 = -12

            if(temp > product){
                // 21 > 18
                product = temp;
            }
        }
        return product; // 21
    }

    public static void main(String[] args) {

        int[] prueba01 = {3, 6, -2, -5, 7, 3};
        int[] prueba02 = {-1, -2};
        int[] prueba03 = {5, 1, 2, 3, 1, 4};
        int[] prueba04 = {1, 2, 3, 0};
        int[] prueba05 = {9, 5, 10, 2, 24, -1, -48};
        int[] prueba06 = {5, 6, -4, 2, 3, 2, -23};
        int[] prueba07 = {4, 1, 2, 3, 1, 5};
        int[] prueba08 = {-23, 4, -3, 8, -12};
        int[] prueba09 = {1, 0, 1, 0, 1000};

        System.out.println(adjacentElementsProductMethod(prueba01)); // 21
        System.out.println(adjacentElementsProductMethod(prueba02)); // 2
        System.out.println(adjacentElementsProductMethod(prueba03)); // 6
        System.out.println(adjacentElementsProductMethod(prueba04)); // 6
        System.out.println(adjacentElementsProductMethod(prueba05)); // 50
        System.out.println(adjacentElementsProductMethod(prueba06)); // 30
        System.out.println(adjacentElementsProductMethod(prueba07)); // 6
        System.out.println(adjacentElementsProductMethod(prueba08)); // -12
        System.out.println(adjacentElementsProductMethod(prueba09)); // 0

    }
}

File: Codesignal-Solutions/Codesignal-Arcade/Level 02/adjacentElementsProduct.cpp
/*

Arcade Level 02 : 04.Adjacent Elements Product

Source Help:
One comment in code fights in this arcade.

Given an array of integers, find the pair of adjacent elements that has the
largest product and return that product

Example:
For inputArray = [3, 6, -2, -5, 7, 3], the output should
be adjacentElementProduct(inputArray) = 21

7 and 3 produce the largest product

Input/Output

[time limit] 500ms
[input]array.integer inputArray

An array of integers containing at least two elements.

Guaranteed constraints:

2 <= inputArray.length <= 10
-1000 <= inputArray[i] <= 1000

[output]

The largest product of adjacent elements.

*/
#include <iostream>
#include <vector>
using namespace std;

int adjacentElementsProduct(vector<int> inputArray){

if ( 2 <= inputArray.size() + 1 <= 10){ // Constraint 1.

    int largestProduct = inputArray.at(0) * inputArray.at(1);
    // Product of first element * second element

    for (int i = 0; i < inputArray.size() - 1; i++){ // Iteration
        if((-1000) <= inputArray.at(i) <= 1000){ // Constraint 2.

            // Condition for each element by right and so on.
            if (inputArray.at(i) * inputArray.at(i+1) > largestProduct){

            // Equality to largestProduct value
            largestProduct = inputArray.at(i) * inputArray.at(i+1);
        }
    }
}
return largestProduct;
}
}

int main(){

cout << adjacentElementsProduct({3, 6, -2, -5, 7, 3});

return 0;
}

